app/forms/questionnaire_form.py

```diff
@@ -37,10 +42,26 @@ def validate(self):
 
         for question in self.schema.get_questions_for_block(self.block_json):
             if question['type'] == 'DateRange':
-                period_from_id = question['answers'][0]['id']
-                period_to_id = question['answers'][1]['id']
-                if not (self.answers_all_valid([period_from_id, period_to_id]) and
-                        self._validate_date_range_question(question['id'], period_from_id, period_to_id)):
+                date_from = question['answers'][0]
+                date_to = question['answers'][1]
+                if self._has_min_and_max_single_dates(date_from, date_to):
+                    # Work out the largest possible range, for date range question
+                    period_range = self._get_period_range_for_single_date(date_from, date_to)
+                    if 'period_limits' in question:
+                        self.validate_date_range_with_period_limits_and_single_date_limits(question['id'],
+                                                                                           question['period_limits'],
+                                                                                           period_range)
+                    else:
+                        self.validate_date_range_with_single_date_limits(question['id'], period_range)
+                period_from_id = date_from['id']
+                period_to_id = date_to['id']
+                messages = None
+                if 'validation' in question:
+                    messages = question['validation'].get('messages')
+                if not (
+                        self.answers_all_valid([period_from_id, period_to_id]) and
+                        self._validate_date_range_question(question['id'], period_from_id, period_to_id, messages,
+                                                           question.get('period_limits'))):
                     valid_form = False
             elif question['type'] == 'Calculated':
                 valid_form = self.validate_calculated_question(question)
```

Comment:
You haven't added this issue in this PR but applying "Boy Scout" rule. If we ever got a block with a date range and a calculation on it (we shouldn't do but I don't think there is anything to stop that possibility) the calculation validation will override the setting of valid_form.